### ISSUE TITLE
from-12-more-optimizations-and-bug-fixes

### ISSUE THREAD

**Issue by david-thrower** [david-thrower/HelixLM#13]
# [Consolidated Dev-Channel] Post-refactor hygiene batch

# Bug fixes to make:

| Item                  | Agent Verdict | My Verdict       | Actual Bug?  | Fix Priority   |
| --------------------- | ------------- | ---------------- | ------------ | -------------- |
| 1. Streaming defeated | High          | **Confirmed**    | Limitation   | P2             |
| 2. Pad/EOS fragility  | High          | **Overcautious** | No (handled) | P2 (assertion) |
| 3. Save race          | High          | **Disagreed**    | No           | None           |
| 4. `_backend` access  | High          | **Confirmed**    | Hygiene      | P1             |
| 5. Warmup leak        | Medium        | **Confirmed**    | Yes (tiny)   | P2             |
| 6. CFG mutation       | Medium        | **Partially**    | Fragile      | P2             |
| 7. Redundant TF32     | Medium        | **Confirmed**    | No           | P3             |
| 8. Dead-code slice    | Medium        | **Confirmed**    | No           | P3             |
| 9. Memory print       | Medium        | **Confirmed**    | No           | P2             |
| 10. No seeding        | Low           | **Confirmed**    | No           | P3             |

You are confused because you have **two different lists merged into one mess**:

1. **The "STOP EVERYTHING" list** — 3 library bugs that corrupt gradients, labels, or GPU memory (P0).
2. **The cofounder agent's list** — 10 code-review nits, some real, some false positives, with ambiguous "High/Medium/Low" labels.

Here is **one single list**, split into unambiguous developer tiers. I have stripped out the false positives and folded the real cofounder findings into their proper priority.

---

## P0 — Ship Blocker (do not train until these are applied)
These four bugs cause the PPL regression, OOM, or LR warmup collapse.

### P0.1 `helix_lm/dataset.py` — EOS tokens masked as padding
**File:** `helix_lm/dataset.py`  
**Problem:** `DocumentAwareDataset.__getitem__` masks every `pad_token_id` in labels as `-100`. Because GPT-2 sets `pad_token = eos_token` (both `50256`), every EOS target is erased. The model never learns to end sequences.

**Fix:** Replace `DocumentAwareDataset.__getitem__` with trailing-pad-only masking.

```python
    def __getitem__(self, idx):
        self._ensure_chunks()
        chunk, is_natural = self.chunks[idx]
        x = torch.tensor(chunk, dtype=torch.long)
        labels = x.clone()

        # Only mask trailing padding we actually appended in _build_chunks
        pad_len = 0
        for i in reversed(range(len(chunk))):
            if chunk[i] == self.pad_id:
                pad_len += 1
            else:
                break
        if pad_len > 0:
            labels[-pad_len:] = -100

        return {
            "input_ids": x,
            "labels": labels,
            "attention_mask": (x != self.pad_id).long(),
            "is_natural_stop": torch.tensor(is_natural, dtype=torch.bool),
        }
```

### P0.2 `helix_lm/dataset.py` — HelixDataset blanket pad mask
**File:** `helix_lm/dataset.py`  
**Problem:** `HelixDataset._make_sample` blanket-masks all `pad_token_id` positions, again colliding with EOS when they share an ID.

**Fix:** Remove the `torch.where` mask. The caller (`__getitem__`) already handles warmup and padding masks exactly.

```python
    def _make_sample(self, chunk, labels, is_natural_stop):
        input_ids = torch.tensor(chunk[:self.seq_len], dtype=torch.long)
        labels_t = torch.tensor(labels[:self.seq_len], dtype=torch.long)
        # DO NOT blanket-mask pad_token_id here — pad_token_id may equal eos_token_id.
        attention_mask = (input_ids != self.tokenizer.pad_token_id).long()
        return {
            "input_ids": input_ids,
            "labels": labels_t,
            "attention_mask": attention_mask,
            "is_natural_stop": torch.tensor(is_natural_stop, dtype=torch.bool),
        }
```

### P0.3 `helix_lm/trainer.py` — Loss logging inflation (**already correct**)
**File:** `helix_lm/trainer.py` (inside `train_epoch`)  
**Problem:** `total_loss += loss.item() * max(1, self.grad_accum_steps)` assumes the last batch was divided by the full `grad_accum_steps`. On the final incomplete accumulation step the actual divisor is smaller, so the printed PPL is artificially inflated.

**Fix:** Track the actual `divisor` and multiply back by it.

```python
            # Scale loss for gradient accumulation
            divisor = 1
            if self.grad_accum_steps > 1:
                is_last = (batch_idx + 1) == len(self.train_loader)
                if is_last and accum_count < self.grad_accum_steps - 1:
                    divisor = accum_count + 1
                else:
                    divisor = self.grad_accum_steps
                loss = loss / divisor

            # ... backward pass ...

            accum_count += 1
            total_loss += loss.item() * divisor   # was: * max(1, self.grad_accum_steps)
            raw_count += 1
```

### 🚧 P0.4 `helix_lm/hf_model.py` — Double-logit OOM (**create a sub-issue** this should be a configurable option)
**File:** `helix_lm/hf_model.py` (inside `HelixForCausalLM.forward`)  
**Problem:** The full `(B, T, V)` `logits` tensor stays in scope while `nn.CrossEntropyLoss` allocates its own large buffers, doubling peak VRAM by ~4.6 GB.

**Fix:** Slice/flatten into `shift_logits`, then delete the parent tensor before CE runs.

```python
        loss = None
        if labels is not None:
            shift_logits = logits[:, :-1, :].reshape(-1, self.config.vocab_size)
            shift_labels = labels[:, 1:].reshape(-1)
            del logits          # free ~4.6 GB before CE allocates
            logits = None
            loss_fct = nn.CrossEntropyLoss(ignore_index=-100)
            loss = loss_fct(shift_logits, shift_labels)
```

### 🛑 P0.5 Training preset — Warmup collapse with gradient accumulation (Out of scope for this codebase)
**File:** your training script preset (e.g., `train_helixlm_lazy.py`)  
**Problem:** If you set `warmup_steps=500` with `grad_accum=4`, the scheduler receives `500 // 4 = 125` optimizer warmup steps. The LR spikes too fast.  
**Fix:** Set `warmup_steps=2000` in the preset. The `Trainer` already divides by `grad_accum_steps`, so the optimizer sees **500** steps.

```python
    256: {
        ...
        "warmup_steps": 2000,   # yields 500 actual optimizer steps when grad_accum=4
        ...
    }
```

---

## P1 — Real Bugs / Material Impact (fix before scaling to 715K)
These will bite you in production or on a 16 GB card, but they do not corrupt training math.

### 🛑 P1.1 Warmup dummy tensor leak (~1.2 GB) (**out of scope for this code base**)
**File:** your training script (`train_helixlm_lazy.py`, startup block)  
**Problem:** A no-grad forward with a dummy batch materializes full logits and never deletes them. This is ~1.2 GB for GPT-2 vocab sitting in CUDA memory at epoch 0.

**Fix:** Delete the dummy input and output immediately.

```python
with torch.no_grad():
    dummy = torch.randint(0, cfg.vocab_size, (cfg.batch_size, cfg.seq_len), device=device)
    _ = model(dummy)
    del dummy, _
    torch.cuda.empty_cache()
```

### P1.2 `tokenizer._backend` accessed directly from script
**File:** `helix_lm/tokenizer.py` + your training script  
**Problem:** Your script calls `tokenizer._backend.save_pretrained(...)`. If `HelixTokenizer` ever refactors its internals, the script crashes.

**Fix:** Add thin public wrappers to `HelixTokenizer`.

```python
# In helix_lm/tokenizer.py
class HelixTokenizer:
    ...
    def save_pretrained(self, path):
        return self._backend.save_pretrained(path)
    def push_to_hub(self, repo_id, **kwargs):
        return self._backend.push_to_hub(repo_id, **kwargs)
```
Then in your script, call `tokenizer.save_pretrained(...)` instead of `tokenizer._backend.save_pretrained(...)`.

### P1.3 Lazy loading defeated by `tqdm(total=...)` **Already correct**
**File:** `helix_lm/trainer.py`  
**Problem:** If `tqdm(..., total=len(self.train_loader))` is present, it forces `Dataset.__len__()` before batch 0, which for `DocumentAwareDataset(lazy=True)` eagerly tokenizes all 715K documents.  
**Fix:** Ensure the tqdm call in `train_epoch` does **not** have a `total=` kwarg.

```python
        pbar = tqdm(
            self.train_loader,
            desc=f"Epoch {epoch}",
            # total=len(self.train_loader),   # REMOVE if present
            unit="batch",
            disable=not self.verbose,
        )
```
*(In the codebase you pasted earlier, this was already absent. Verify your local file.)*

### P1.4 🛑 Assert `pad_token_id` is not None before training **Out of scope for this code base**
**File:** your training script  
**Problem:** If a tokenizer backend is ever added without the `pad_token = eos_token` backfill, training crashes with `NoneType` inside the dataset.  
**Fix:** One-line guard in the training script.

```python
assert tokenizer.pad_token_id is not None, "Tokenizer must have a pad_token_id set"
cfg.pad_token_id = tokenizer.pad_token_id
cfg.eos_token_id = tokenizer.eos_token_id
cfg.bos_token_id = tokenizer.bos_token_id
```

### P1.5 🛑 Memory print is misleading  **Out of scope for this code base**
**File:** your training script  
**Problem:** `torch.cuda.memory_allocated()` does not show allocator overhead / fragmentation. You need `memory_reserved()` to diagnose OOM.  
**Fix:** Print both.

```python
print(f"VRAM allocated: {torch.cuda.memory_allocated()/1e9:.2f} GB")
print(f"VRAM reserved:  {torch.cuda.memory_reserved()/1e9:.2f} GB")
```

---

### P2.1 Stop mutating `cfg` between training stages
**File:** your training script  
**Problem:** You mutate `cfg.lr`, `cfg.epochs`, etc. between stage-1 and stage-2 Trainer construction. It works today because you mutate before construction, but it is fragile.

**Fix:** Pass explicit kwargs to `Trainer.__init__`.

```python
# Instead of:
cfg.lr = 1e-3
trainer = Trainer(model, cfg, ...)

# Prefer:
trainer = Trainer(
    model=model, cfg=cfg, train_texts=...,
    lr=1e-3, epochs=3, batch_size=24, ...
)
```
*(This requires adding `lr`, `epochs`, `batch_size` as optional kwargs to `Trainer.__init__` that override `cfg`.)*

### P2.2 Add seeding for reproducibility
**File:** your training script  
**Fix:**

```python
import random, numpy as np, torch
SEED = 42
random.seed(SEED)
np.random.seed(SEED)
torch.manual_seed(SEED)
# torch.use_deterministic_algorithms(True)  # optional, may slow CUDA
```

### P2.3 Streaming design debt
**File:** your training script + `helix_lm/trainer.py`  
**Problem:** `load_texts(..., streaming=True)` still appends to a Python `list`, so all 715K documents land in host RAM.  
**Fix:** This is an architectural refactor. For now, cap `max_samples` to fit in RAM. A true streaming fix requires `Trainer` to accept a generator and `DocumentAwareDataset` to chunk on-the-fly without `__len__`.

---

### P3.1 Remove redundant TF32 lines
**File:** your training script  
`torch.set_float32_matmul_precision("highest")` already disables TF32. The two explicit `allow_tf32 = False` lines are dead weight. Delete them.

### P3.2 Remove dead-code slice
**File:** your training script  
`pretrain_train = pretrain_train[:args.max_samples]` after `load_texts` already respected `max_samples`. Delete the redundant slice.

---

## Skip — False Positive (do nothing)

- **Save directory race** (cofounder item 3): HF `save_pretrained` already calls `os.makedirs(..., exist_ok=True)`. There is no race.

---

## Your Exact Run Order

Apply **only P0** first, then run:

```bash
export PYTORCH_CUDA_ALLOC_CONF="expandable_segments:True,max_split_size_mb:512"

python train_helixlm_lazy.py \
  --seq-len 256 \
  --max-samples 10000 \
  --output-dir ./helixlm-fixed-eos \
  --hf-org david-thrower
```

If epoch-1 PPL is **< 250 and trending down**, you are fixed. Do not touch P1–P3 until that smoke test passes.

**Comment by david-thrower** [comment]
## Final Pre-Flight Checklist

---

### ✅ DONE — Fixed in library code (no action needed)

| Fix | File | Evidence |
|-----|------|----------|
| **EOS/pad collision** (`DocumentAwareDataset`) | `dataset.py` | Trailing-pad-only masking via reversed scan. |
| **EOS/pad collision** (`HelixDataset._make_sample`) | `dataset.py` | Blanket `torch.where` removed; caller handles exact masks. |
| **Loss accumulation inflation** | `trainer.py` | `total_loss += loss.item() * divisor` (not `grad_accum_steps`). |
| **Missing `.backward()`** | `trainer.py` | Present after scaling with proper AMP branching. |
| **Lazy loading defeated** | `trainer.py` | Scheduler init deferred from `__init__` to `train_epoch`. |
| **Double-logit OOM gate** | `hf_model.py` | `del logits` wrapped behind `memory_efficient_forward` flag. |
| **Config flag** | `config.py` | `memory_efficient_forward: bool = False` declared and stored. |
| **Tokenizer public API** | `tokenizer.py` | `save_pretrained()` and `push_to_hub()` wrappers exist. |

---

### ⚠️ NOT in this repo's scope: You must handle these in the training script / environment

These are **not in the library**; they are preset / CLI / environment choices you must make correctly before launch.

| Item | What you must do | Why it matters |
|------|------------------|----------------|
| **Warmup steps scaling** | Set `"warmup_steps": 2000` in your preset (for `grad_accum=4`). | Trainer divides by `grad_accum_steps`. 2000 → 500 actual optimizer warmup steps. 500 → 125 (too sharp). |
| **OOM flag activation** | Set `memory_efficient_forward=True` in your `HelixConfig` call if you are VRAM-constrained. | Default is `False` (HF-compatible). For 16 GB with batch=24, you may not need it, but if you increase batch size later, you will. |
| **CUDA allocator** | `export PYTORCH_CUDA_ALLOC_CONF="expandable_segments:True,max_split_size_mb:512"` before running. | Prevents fragmentation OOMs on 16 GB cards. |
| **A4000-safe preset** | Use `batch_size=24`, `grad_accum=4`, `seq_len=256` for 16 GB. | Effective batch = 96. Fits safely with headroom. |

---

### 📋 NOT DONE — Lower priority (Deferred to next sprint , not blocking, codebase is stable pending the GPU runs and addition of a long - text CICD test)

These are real but will not corrupt your training run. Move to the next dev-channel sprint.

| Item | Type | Notes |
|------|------|-------|
| Add a beyond sequence length CICD test | CICD |**required** |
| Warmup dummy tensor leak | Script hygiene | `del dummy, _; torch.cuda.empty_cache()` after warmup forward. |
| Assert `pad_token_id` | Defensive | One-line guard before `Trainer` construction. |
| Memory print | Diagnostic | Print both `allocated` and `reserved`. |
| Stop mutating `cfg` | Architecture | Pass explicit kwargs to `Trainer.__init__`. |
| Deterministic seeding | Reproducibility | Add `--seed` CLI flag. |
| Streaming design debt | Architectural | `load_texts` materializes full list. Cap `max_samples` for now. |
| Redundant TF32 lines | Cosmetic | Delete explicit `allow_tf32=False` lines. |
| Dead-code slice | Cosmetic | Remove redundant `[:max_samples]` slice. |
| README `_backend` example | Documentation | README still shows `tokenizer._backend.save_pretrained()`. Update to public API. |

---

### 🔍 KNOWN INEFFICIENCY (not a bug, just waste)

**Dead `self.model.head` in `HelixForCausalLM`**
- `HelixLMCore` instantiates `self.head` (un-tied, since `HelixForCausalLM` passes `tie_weights=False`).
- `HelixForCausalLM` then creates its own `self.lm_head` and uses that exclusively.
- Result: `self.model.head` (~6.4M params for GPT-2 vocab) is never used but still holds optimizer state (~51MB of AdamW buffers).
- **Impact:** Wastes ~30% of parameter count and a small amount of optimizer memory. Does not affect correctness.
- **Fix later:** Pass `tie_weights=True` to `HelixLMCore` and reuse `self.model.head`, or delete it in `HelixForCausalLM.__init__`.

**Comment by david-thrower** [comment]
Closed in #9
